### PR TITLE
[feature/semantic_text] Register semantic text sub fields in the mapping

### PR DESCRIPTION
**Pull request by jimczi**
This PR refactors the semantic text field mapper to register its sub fields in the mapping instead of re-creating them each time when parsing documents.
It also fixes the generation of these fields in case the semantic text field is defined in an object field.
Lastly this change adds a new section called `model_settings` in the field parameter that is updated by the field mapper when inference results are received from a bulk action. The model settings are available in the fields as soon as the first document with the inference field is ingested and they are used to validate that updates. They are used to ensure consistency between what's used in the bulk action and what's defined in the field mapping.

Note: This PR is opened against the feature branch: feature/semantic_text

### CODE REVIEW

**Review by carlosdelest — APPROVED** [review]
LGTM, but I'd like to check if we can avoid serializing `inference_id` on model settings for the field mapper, to avoid duplicating that information in the mapping.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:78]: It seems a bit silly that we're defining the index analyzers, given that we're not indexing or storing the text field.

The point of this text field is to store the text that generated the chunk in `_source`, correct? What if we made it a `keyword` field instead?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java:165]: Doesn't a valid model settings also require an inference ID?

**Review by Mikep86 — APPROVED** [review]
Left some comments, none that I think block merging though

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java:304]: This PR preserves this behaviour but only for the `_inference.$field_name.results` object. I am not a fan of this leniency but it is needed since we'll want to extend the information we provide in the future. I thought that we can be strict on the top level fields here since they should not change.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java:348]: Right, although this PR doesn't change the behaviour, it was already the case before. I think we can work on adding a post parsing validation in the metadata field mapper in a follow up.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:78]: Yep good call, I wonder if we need to explicitly map this field though. Maybe we can just avoid creating it in the mapping entirely?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java:165]: It's already checked in the constructor, inferenceId cannot be null.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java:242]: I don't think it's an issue, we do parse the results twice but we do it per field so it won't take a lot of memory.
So unless we change the structure of the object we'll need to do this double parsing since we cannot rely on the order of the field in a json map.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java
Comment (Mikep86) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java:165]: Got it, didn't see it initially because that code didn't change

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (Mikep86) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:78]: @carlosdelest WDYT?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java
Comment (Mikep86) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java:165]: Nitpick: Since this method is only about validating the task type, what if we:

- Renamed it to `validateTaskType`
- Used it to check if `taskType` is null
- Called it from `SemanticTextModelSettings(TaskType taskType, String inferenceId, Integer dimensions, SimilarityMeasure similarity)` so we always check if the task type is valid

Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java:166]: That's the intent. The validate call was put on the wrong constructor so I fixed that but taskType is already required to be non-null. I also added more validations so I think the naming is fine.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java:357]: I ended up adding a fix in this PR. In the bulk action we now remove the `_inference` field if none of the inference fields are present so I think we're covered there.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:180]: Correct. This is also needed to ensure that we're not updating the sub-fields with non-compatible changes (like changing options in the field types).

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java
Comment (jimczi) [reply x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java:153]: Yep good call, I removed the inference_id from the model_settings and added a new top level section in the _inference object to set it in https://github.com/elastic/elasticsearch/pull/106560/commits/2be50d7338083a51f11c57e166d2412d56daeadb. Let me know what you think.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (carlosdelest) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:78]: I don't think we want to index or analyze it in any way; users can use multifields / copy_to in case they want to search for it. I'd vote for skipping the mapping.

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/20_semantic_text_field_mapper.yml
Comment (carlosdelest) [inline x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/20_semantic_text_field_mapper.yml:58]: These tests were meant to address the parsing of the inference results. I don't see this explicitly tested in other tests. Should we keep these?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (carlosdelest) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:55]: We're not using the logger anymore I think

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java
Comment (carlosdelest) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java:243]:
```suggestion
```

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java
Comment (carlosdelest) [reply x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java:153]: That will work 👍

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java]: I switched to a keyword field instead so that it's clear that the field is there but not indexed.

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/20_semantic_text_field_mapper.yml
Comment (jimczi) [reply x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/20_semantic_text_field_mapper.yml:58]: I moved these tests to InferenceMetadataFieldMapperTests since it is not possible anymore to provide the _inference field manually in a bulk request.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java]: https://github.com/elastic/elasticsearch/pull/106560/commits/b3fb5d3017454e668c6e2249c425caa3bd5a0ec4

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java
Comment (Mikep86) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextModelSettings.java:161]: String formatting error: There's no space between "TEXT_EMBEDDING" and "or"

**Review by Mikep86 — APPROVED** [review]

**Pull request merged**
